<a href="#">
    <img alt="Caixa SINAPI logo" src="./img/sinapi_logo.jpg" style="height:140px; align:left">
</a>

# Composições de custos analíticas

<h1>Relatório</h1>

O presente relatório foi construído a partir de 3 fontes de composições analíticas. A primeira delas consiste na [SINAPI DEZ/2020](https://www.caixa.gov.br/poder-publico/modernizacao-gestao/sinapi/referencias-precos-insumos/Paginas/default.aspx). A segunda fonte se refere aos dados publicados pela tabela de referência [GOINFRA NOV/2020](http://www.goinfra.go.gov.br/Tabela-de-Composicao/114). Por fim, também foram inseridas composições elaboradas pela [Secretaria de Estado da Educação (SEE-GO)](http://seduc.go.gov.br/). Todos os preços são DESONERADOS e referentes ao estado de Goiás.

<h2>Sumário</h2>

1) [Introdução](#intro)
2) [Análise exploratória dos dados](#data_exploration)<br>
a. [Qual a estrutura dos datasets?](#question_1)<br>
b. [Há inconsistências nos dados disponíveis?](#question_2)<br>
c. [Quais as principais características das variáveis?](#question_3)<br>
d. [Qual a frequência de itens por ano?](#question_4)<br>
3) [Tratamento de dados](#data_cleaning)<br>
a. [Eliminação de dados ausentes](#handle_missing)<br>
b. [Eliminação de dados dupilcados](#handle_duplicate)<br>
c. [Substituição de anos incorretos](#wrong_year)<br>
4) [Visualização](#data_visualization)<br>
a. [Jogos mais vendidos](#top_sales)<br>
b. [Total de vendas](#total_sales)<br>
c. [Número de vendas por gênero](#genre_sales)<br>
d. [Editoras mais populares no top 100](#top_publishers)<br>
5) [Conclusão](#conclusion)

# 2 Análise Exploratória <a id="data_exploration"></a>

#### Importação de bibliotecas

Dada sua elevada flexibilidade e capacidade de processamento, optou-se por analisar os dados com a biblioteca pandas.

In [1]:
import pandas as pd
from utils import helpers

!pip install xlrd

#### Arquivos fonte

A cada atualização publicada pelo órgãos, os arquivos devem ser substituídos na pasta app/data.

In [2]:
goinfra_sintetica_filepath = "./data/goinfra-composicoes-sinteticas-11.2020.xls"
goinfra_analitica_filepath = "./data/goinfra-composicoes-analiticas-11.2020.xls"
sinapi_analitica_filepath = "./data/sinapi-composicoes-analiticas-12.2020.xls"

### 2.1 Qual a estrutura dos datasets?<a id="question_1"></a>

#### a. Composições sintéticas GOINFRA

Devido à estrutura do arquivo em excel, as 6 primeiras linhas são desconsideradas.

In [3]:
goinfra_sintetica = pd.read_excel(goinfra_sintetica_filepath, skiprows=6)
goinfra_sintetica

CODIGO DA COMPOSICAO  \
0                                     164    
1                                  020000    
2                                  020100    
3                                  020101    
4                                  020102    
...                                    ...   
2031                               271851    
2032                               271852    
2033                               271853    
2034                                   NaN   
2035  SCO - Sistema de Custos e Orçamentos   

                                DESCRICAO DA COMPOSICAO  Unnamed: 2  \
0                                 SERVIÇOS PRELIMINARES         NaN   
1                                 SERVIÇOS PRELIMINARES         NaN   
2     DEMOLIÇÃO MANUAL - COBERTURA TELHA METÁLICA C/...         NaN   
3     DEMOLICAO MANUAL COBERTURA TELHA CERAMICA C/ T...         NaN   
4     DEMOLICAO MANUAL COBERTURA TELHA FIBROCIMENTO/...         NaN   
...                                                 ...         ...   
2031                          LETRA CAIXA INOX COLOCADA         NaN   
2032                 LETRA CAIXA INOX ESCOVADO COLOCADA         NaN   
2033                LETRA  CAIXA LATAO AMARELO COLOCADA         NaN   
2034                                                NaN         NaN   
2035                                                NaN         NaN   

      Unnamed: 3  Unnamed: 4 Unnamed: 5 UNIDADE  Unnamed: 7  CUSTO MATERIAL  \
0            NaN         NaN        NaN     NaN         NaN             NaN   
1            NaN         NaN        NaN  S/U            NaN            0.00   
2            NaN         NaN        NaN  m2             NaN            0.00   
3            NaN         NaN        NaN  m2             NaN            0.00   
4            NaN         NaN        NaN  m2             NaN            0.00   
...          ...         ...        ...     ...         ...             ...   
2031         NaN         NaN        NaN  m              NaN          377.93   
2032         NaN         NaN        NaN  m              NaN          464.31   
2033         NaN         NaN        NaN  m              NaN          583.09   
2034         NaN         NaN        NaN     NaN         NaN             NaN   
2035         NaN         NaN     AGETOP     NaN         NaN             NaN   

      CUSTO MAO DE OBRA  Unnamed: 10  CUSTO TOTAL  
0                   NaN          NaN          NaN  
1                  0.00          NaN         0.00  
2                  2.33          NaN         2.33  
3                  3.73          NaN         3.73  
4                  2.07          NaN         2.07  
...                 ...          ...          ...  
2031               0.00          NaN       377.93  
2032               0.00          NaN       464.31  
2033               0.00          NaN       583.09  
2034                NaN          NaN          NaN  
2035                NaN          NaN          NaN  

[2036 rows x 12 columns]

Há 2036 linhas e 12 colunas.

#### b. Composições analíticas GOINFRA

A leitura do arquivo de composições analíticas da GOINFRA é apresentado a seguir.

In [4]:
goinfra_analitica = pd.read_excel(goinfra_analitica_filepath)
goinfra_analitica

CODIGO DA COMPOSICAO CODIGO ITEM  \
0                    20100       0004    
1                    20100       0005    
2                    20101       0005    
3                    20101       0004    
4                    20102       0004    
...                    ...         ...   
9482                271803       2500    
9483                271850       2671    
9484                271851       2670    
9485                271852       2683    
9486                271853       2684    

                                         DESCRIÇÃO ITEM UNIDADE ITEM  \
0                                              PEDREIRO            h   
1                                              SERVENTE            h   
2                                              SERVENTE            h   
3                                              PEDREIRO            h   
4                                              PEDREIRO            h   
...                                                 ...          ...   
9482                     LADRILHO HIDRAULICO DUAS CORES       m2       
9483  LETRA CX.CHAPA GALV.PINT.COLOCADA (ALT.LETRAXN...       m        
9484  LETRA CAIXA INOX  COLOCADA (ALT.LETRASXNUMERO ...       m        
9485  LETRA CX.INOX ESCOVADO COLOCADA(ALTURA LETRASX...       m        
9486  LETRA CX.LATAO AMARELO COLOCADA(ALTURA LETRASX...       m        

      COEFICIENTE  CUSTO TOTAL.1  
0          0.0225           0.32  
1          0.2250           2.01  
2          0.3600           3.21  
3          0.0360           0.52  
4          0.0200           0.29  
...           ...            ...  
9482       1.0500          53.16  
9483       1.0000         313.14  
9484       1.0000         377.93  
9485       1.0000         464.31  
9486       1.0000         583.09  

[9487 rows x 6 columns]

Observam-se 9487 linhas e 6 colunas. A coluna 'CODIGO DA COMPOSICAO' será usada posteriormente como chave para busca de composições analíticas.

#### c. Composições analíticas SINAPI

As colunas se iniciam a partir da 6ª linha, logo as 5 primeiras são desconsideradas.

In [5]:
sinapi_analitica = pd.read_excel(sinapi_analitica_filepath, skiprows=5)
sinapi_analitica.head()

DESCRICAO DA CLASSE SIGLA DA CLASSE  \
0                            NaN             NaN   
1  ASSENTAMENTO DE TUBOS E PECAS            ASTU   
2  ASSENTAMENTO DE TUBOS E PECAS            ASTU   
3  ASSENTAMENTO DE TUBOS E PECAS            ASTU   
4  ASSENTAMENTO DE TUBOS E PECAS            ASTU   

                                 DESCRICAO DO TIPO 1  SIGLA DO TIPO 1  \
0                                                NaN              NaN   
1  FORNEC E/OU ASSENT DE TUBO DE FERRO FUNDIDO JUNTA             45.0   
2  FORNEC E/OU ASSENT DE TUBO DE FERRO FUNDIDO JUNTA             45.0   
3  FORNEC E/OU ASSENT DE TUBO DE FERRO FUNDIDO JUNTA             45.0   
4  FORNEC E/OU ASSENT DE TUBO DE FERRO FUNDIDO JUNTA             45.0   

   CODIGO DO AGRUPADOR DESCRICAO DO AGRUPADOR CODIGO DA COMPOSICAO  \
0                  NaN                    NaN                  NaN   
1                  NaN                    NaN                97141   
2                  NaN                    NaN                97141   
3                  NaN                    NaN                97141   
4                  NaN                    NaN                97141   

                             DESCRICAO DA COMPOSICAO UNIDADE  \
0                                                NaN     NaN   
1  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...       M   
2  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...       M   
3  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...       M   
4  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...       M   

       ORIGEM DE PREÇO  ... CUSTO MATERIAL  % MATERIAL  CUSTO EQUIPAMENTO  \
0                  NaN  ...            NaN         NaN                NaN   
1  ATRIBUÍDO SÃO PAULO  ...           1,28  18,7406296               1,15   
2  ATRIBUÍDO SÃO PAULO  ...            NaN         NaN                NaN   
3  ATRIBUÍDO SÃO PAULO  ...            NaN         NaN                NaN   
4  ATRIBUÍDO SÃO PAULO  ...            NaN         NaN                NaN   

  % EQUIPAMENTO CUSTO SERVICOS TERCEIROS % SERVICOS TERCEIROS CUSTO OUTROS  \
0           NaN                      NaN                  NaN          NaN   
1    16,9415292                     0,00            0,0000000         0,00   
2           NaN                      NaN                  NaN          NaN   
3           NaN                      NaN                  NaN          NaN   
4           NaN                      NaN                  NaN          NaN   

    % OUTROS % ATRIBUÍDO SÃO PAULO            VINCULO  
0        NaN                   NaN                NaN  
1  0,0000000               0.16667  CAIXA REFERENCIAL  
2        NaN                   NaN  CAIXA REFERENCIAL  
3        NaN                   NaN  CAIXA REFERENCIAL  
4        NaN                   NaN  CAIXA REFERENCIAL  

[5 rows x 31 columns]

In [6]:
sinapi_analitica.shape

(42052, 31)

Há um total de 42.052 linhas e 31 colunas, cujos nomes são apresentados abaixo.

In [7]:
sinapi_analitica.columns.values

array(['DESCRICAO DA CLASSE', 'SIGLA DA CLASSE', 'DESCRICAO DO TIPO 1',
       'SIGLA DO TIPO 1', 'CODIGO DO AGRUPADOR', 'DESCRICAO DO AGRUPADOR',
       'CODIGO DA COMPOSICAO', 'DESCRICAO DA COMPOSICAO', 'UNIDADE',
       'ORIGEM DE PREÇO', 'CUSTO TOTAL', 'TIPO ITEM', 'CODIGO ITEM',
       'DESCRIÇÃO ITEM', 'UNIDADE ITEM', 'ORIGEM DE PREÇO ITEM',
       'COEFICIENTE', 'PRECO UNITARIO', 'CUSTO TOTAL.1',
       'CUSTO MAO DE OBRA', '% MAO DE OBRA', 'CUSTO MATERIAL',
       '% MATERIAL', 'CUSTO EQUIPAMENTO', '% EQUIPAMENTO',
       'CUSTO SERVICOS TERCEIROS', '% SERVICOS TERCEIROS', 'CUSTO OUTROS',
       '% OUTROS', '% ATRIBUÍDO SÃO PAULO', 'VINCULO'], dtype=object)

Nem todas colunas são relevantes para a presente análise. Apenas as colunas referentes à códigos, descrições e preços serão mantidas.

### 2.2 Há inconsistências no conjunto de dados?<a id="question_2"></a>

#### a. Composições sintéticas GOINFRA

##### Dados nulos

Procedeu-se à investigação quanto à quantidade de dados nulos no conjunto de dados.

In [8]:
goinfra_sintetica.isna().sum()

CODIGO DA COMPOSICAO          1
DESCRICAO DA COMPOSICAO       2
Unnamed: 2                 2036
Unnamed: 3                 2036
Unnamed: 4                 2036
Unnamed: 5                 2035
UNIDADE                      89
Unnamed: 7                 2036
CUSTO MATERIAL               28
CUSTO MAO DE OBRA            28
Unnamed: 10                2036
CUSTO TOTAL                  28
dtype: int64

Observa-se uma quantidade significativa de colunas vazias. Isso se deve ao fato de que o arquivo de excel apresenta mesclagem de diversas colunas. Desse modo, as colunas cujo nome contenha 'Unnamed' serão removidas.

A tabela abaixo apresenta as linhas com valores nulos para a coluna 'CUSTO TOTAL'.

In [9]:
goinfra_sintetica[goinfra_sintetica['UNIDADE'].isna()].reset_index()

index                  CODIGO DA COMPOSICAO DESCRICAO DA COMPOSICAO  \
0       0                                  164    SERVIÇOS PRELIMINARES   
1      83                                  165              TRANSPORTES   
2      84                               030000              TRANSPORTES   
3      94                                  166         SERVIÇO EM TERRA   
4      95                               040000         SERVICO EM TERRA   
..    ...                                   ...                     ...   
84   1889                               260000                  PINTURA   
85   1941                                  189                 DIVERSOS   
86   1942                               270000                 DIVERSOS   
87   2034                                   NaN                     NaN   
88   2035  SCO - Sistema de Custos e Orçamentos                     NaN   

    Unnamed: 2  Unnamed: 3  Unnamed: 4 Unnamed: 5 UNIDADE  Unnamed: 7  \
0          NaN         NaN         NaN        NaN     NaN         NaN   
1          NaN         NaN         NaN        NaN     NaN         NaN   
2          NaN         NaN         NaN        NaN     NaN         NaN   
3          NaN         NaN         NaN        NaN     NaN         NaN   
4          NaN         NaN         NaN        NaN     NaN         NaN   
..         ...         ...         ...        ...     ...         ...   
84         NaN         NaN         NaN        NaN     NaN         NaN   
85         NaN         NaN         NaN        NaN     NaN         NaN   
86         NaN         NaN         NaN        NaN     NaN         NaN   
87         NaN         NaN         NaN        NaN     NaN         NaN   
88         NaN         NaN         NaN     AGETOP     NaN         NaN   

    CUSTO MATERIAL  CUSTO MAO DE OBRA  Unnamed: 10  CUSTO TOTAL  
0              NaN                NaN          NaN          NaN  
1              NaN                NaN          NaN          NaN  
2              0.0                0.0          NaN          0.0  
3              NaN                NaN          NaN          NaN  
4              0.0                0.0          NaN          0.0  
..             ...                ...          ...          ...  
84             0.0                0.0          NaN          0.0  
85             NaN                NaN          NaN          NaN  
86             0.0                0.0          NaN          0.0  
87             NaN                NaN          NaN          NaN  
88             NaN                NaN          NaN          NaN  

[89 rows x 13 columns]

O número de linhas com valor nulo coincide com o encontrado anteriormente para a coluna 'UNIDADE'. As linhas em que essa coluna tenha valor nulo serão removidas durante o tratamento de dados. É razoável supor que linhas com valores nulos nas colunas 'CODIGO DA COMPOSICAO', 'DESCRICAO DA COMPOSICAO', 'CUSTO MATERIAL', 'CUSTO MAO DE OBRA' e 'CUSTO TOTAL' também serão removidas com esse procedimento.

##### Tipos de variáveis

Os tipos de cada variável são apresentados a seguir.

In [10]:
goinfra_sintetica.dtypes

CODIGO DA COMPOSICAO        object
DESCRICAO DA COMPOSICAO     object
Unnamed: 2                 float64
Unnamed: 3                 float64
Unnamed: 4                 float64
Unnamed: 5                  object
UNIDADE                     object
Unnamed: 7                 float64
CUSTO MATERIAL             float64
CUSTO MAO DE OBRA          float64
Unnamed: 10                float64
CUSTO TOTAL                float64
dtype: object

É sabido que os códigos de composições sintéticas são estritamente numéricos. Logo, há necessidade de convertê-los para int.

#### b. Composições analíticas GOINFRA

##### Dados nulos

Procedeu-se à investigação quanto à presença de valores nulos nos dados em questão.

In [11]:
goinfra_analitica.describe()

CODIGO DA COMPOSICAO  COEFICIENTE  CUSTO TOTAL.1
count           9487.000000  9487.000000    9487.000000
mean          120916.831243     5.277379      57.832741
std            77220.560709    46.247697     699.643946
min            20100.000000     0.000100       0.000000
25%            71145.500000     0.130750       1.180000
50%            81571.000000     0.530000       3.920000
75%           180380.000000     1.020000      16.655000
max           271853.000000  1817.090000   40676.070000

In [12]:
goinfra_analitica.isna().sum()

CODIGO DA COMPOSICAO    0
CODIGO ITEM             0
DESCRIÇÃO ITEM          0
UNIDADE ITEM            0
COEFICIENTE             0
CUSTO TOTAL.1           0
dtype: int64

Não há valores nulos no arquivo, logo não há necessidade de eliminação de linhas.

##### Tipos de variáveis

Os tipos de cada variável são apresentados abaixo.

In [13]:
goinfra_analitica.dtypes

CODIGO DA COMPOSICAO      int64
CODIGO ITEM              object
DESCRIÇÃO ITEM           object
UNIDADE ITEM             object
COEFICIENTE             float64
CUSTO TOTAL.1           float64
dtype: object

Pode-se dizer que os tipos são compatíveis com o esperado para cada variável. Os códigos das composições são numéricos. Enquanto isso, o código de item pode se mostrar na forma de texto (ex.: H689) o que justifica o tipo 'object'.

#### c. Composições analíticas SINAPI

##### Dados nulos

As quantidades de valores nulos presentes no conjunto de dados são apresentadas abaixo.

In [14]:
sinapi_analitica.isna().sum()

DESCRICAO DA CLASSE             1
SIGLA DA CLASSE                 1
DESCRICAO DO TIPO 1             1
SIGLA DO TIPO 1                 1
CODIGO DO AGRUPADOR         41893
DESCRICAO DO AGRUPADOR      41893
CODIGO DA COMPOSICAO            1
DESCRICAO DA COMPOSICAO         1
UNIDADE                         1
ORIGEM DE PREÇO                 1
CUSTO TOTAL                     1
TIPO ITEM                    6814
CODIGO ITEM                  6814
DESCRIÇÃO ITEM               6814
UNIDADE ITEM                 6814
ORIGEM DE PREÇO ITEM         6814
COEFICIENTE                  6814
PRECO UNITARIO               6814
CUSTO TOTAL.1                6814
CUSTO MAO DE OBRA           35239
% MAO DE OBRA               35239
CUSTO MATERIAL              35239
% MATERIAL                  35239
CUSTO EQUIPAMENTO           35239
% EQUIPAMENTO               35239
CUSTO SERVICOS TERCEIROS    35239
% SERVICOS TERCEIROS        35239
CUSTO OUTROS                35239
% OUTROS                    35239
% ATRIBUÍDO SÃ

Pode-se presumir que a elevada quantidade de valores nulos se deva ao fato de que as composições sintéticas e analíticas são dispostas no mesmo arquivo. Desse modo, faz-se necessário um tratamento de dados. Em primeiro lugar, deve-se obter a lista de composições sintéticas a partir da eliminação de duplicatas. Além disso, apenas algumas colunas devem ser extraídas, tanto para a tabela sintética quanto a analítica.

##### Tipos de variáveis

Os tipos de cada variável são apresentados a seguir.

In [15]:
sinapi_analitica.dtypes

DESCRICAO DA CLASSE          object
SIGLA DA CLASSE              object
DESCRICAO DO TIPO 1          object
SIGLA DO TIPO 1             float64
CODIGO DO AGRUPADOR         float64
DESCRICAO DO AGRUPADOR       object
CODIGO DA COMPOSICAO         object
DESCRICAO DA COMPOSICAO      object
UNIDADE                      object
ORIGEM DE PREÇO              object
CUSTO TOTAL                  object
TIPO ITEM                    object
CODIGO ITEM                 float64
DESCRIÇÃO ITEM               object
UNIDADE ITEM                 object
ORIGEM DE PREÇO ITEM         object
COEFICIENTE                  object
PRECO UNITARIO               object
CUSTO TOTAL.1                object
CUSTO MAO DE OBRA            object
% MAO DE OBRA                object
CUSTO MATERIAL               object
% MATERIAL                   object
CUSTO EQUIPAMENTO            object
% EQUIPAMENTO                object
CUSTO SERVICOS TERCEIROS     object
% SERVICOS TERCEIROS         object
CUSTO OUTROS                

Vale ressaltar que o código da composição é do tipo objeto, tendo em vista que podem existir valores na forma de texto (ex. 74291/1). A coluna 'CODIGO ITEM' deveria ser do tipo int. Por fim, as colunas de custo deveriam ser do tipo float.

# 3. Tratamento de dados <a id="data_cleaning"></a>

### a. Composições sintéticas GOINFRA <a id="handle_goinfra_sintetica"></a>

#### Eliminação de dados nulos

Primeiramente, foi efetuada a eliminação das colunas que contenham o text 'Unnamed' em seu título.

In [16]:
goinfra_sintetica.drop(goinfra_sintetica.columns[goinfra_sintetica.columns.str.contains('Unnamed',case = False)],axis = 1, inplace = True)
goinfra_sintetica.reset_index(drop=True, inplace=True)
goinfra_sintetica

CODIGO DA COMPOSICAO  \
0                                     164    
1                                  020000    
2                                  020100    
3                                  020101    
4                                  020102    
...                                    ...   
2031                               271851    
2032                               271852    
2033                               271853    
2034                                   NaN   
2035  SCO - Sistema de Custos e Orçamentos   

                                DESCRICAO DA COMPOSICAO UNIDADE  \
0                                 SERVIÇOS PRELIMINARES     NaN   
1                                 SERVIÇOS PRELIMINARES  S/U      
2     DEMOLIÇÃO MANUAL - COBERTURA TELHA METÁLICA C/...  m2       
3     DEMOLICAO MANUAL COBERTURA TELHA CERAMICA C/ T...  m2       
4     DEMOLICAO MANUAL COBERTURA TELHA FIBROCIMENTO/...  m2       
...                                                 ...     ...   
2031                          LETRA CAIXA INOX COLOCADA  m        
2032                 LETRA CAIXA INOX ESCOVADO COLOCADA  m        
2033                LETRA  CAIXA LATAO AMARELO COLOCADA  m        
2034                                                NaN     NaN   
2035                                                NaN     NaN   

      CUSTO MATERIAL  CUSTO MAO DE OBRA  CUSTO TOTAL  
0                NaN                NaN          NaN  
1               0.00               0.00         0.00  
2               0.00               2.33         2.33  
3               0.00               3.73         3.73  
4               0.00               2.07         2.07  
...              ...                ...          ...  
2031          377.93               0.00       377.93  
2032          464.31               0.00       464.31  
2033          583.09               0.00       583.09  
2034             NaN                NaN          NaN  
2035             NaN                NaN          NaN  

[2036 rows x 6 columns]

No entanto, ainda há valores nulos a serem eliminados.

In [17]:
goinfra_sintetica.isna().sum()

CODIGO DA COMPOSICAO        1
DESCRICAO DA COMPOSICAO     2
UNIDADE                    89
CUSTO MATERIAL             28
CUSTO MAO DE OBRA          28
CUSTO TOTAL                28
dtype: int64

Tendo em vista que a coluna unidade possui a maior quantidade de linhas com valores nulos, optou-se por eliminá-las.

In [18]:
goinfra_sintetica = goinfra_sintetica.dropna(subset=['UNIDADE'])

A seguir, é feita uma nova verificação de valores nulos.

In [19]:
goinfra_sintetica.isna().sum()

CODIGO DA COMPOSICAO       0
DESCRICAO DA COMPOSICAO    0
UNIDADE                    0
CUSTO MATERIAL             0
CUSTO MAO DE OBRA          0
CUSTO TOTAL                0
dtype: int64

É possível perceber que todas as linhas com valores nulos foram devidamente tratadas.

#### Conversão de tipo

A coluna de código deve ser convertida para int.

In [20]:
goinfra_sintetica['CODIGO DA COMPOSICAO'].astype(str).astype(int)

1        20000
2        20100
3        20101
4        20102
5        20103
         ...  
2029    271803
2030    271850
2031    271851
2032    271852
2033    271853
Name: CODIGO DA COMPOSICAO, Length: 1947, dtype: int64

In [21]:
goinfra_sintetica.dtypes

CODIGO DA COMPOSICAO        object
DESCRICAO DA COMPOSICAO     object
UNIDADE                     object
CUSTO MATERIAL             float64
CUSTO MAO DE OBRA          float64
CUSTO TOTAL                float64
dtype: object

Os dados acima indicam que a conversão foi bem sucedida.

### b. Composições analíticas GOINFRA <a id="handle_goinfra_analitica"></a>

Conforme indicado anteriormente, não há necessidade de efetuar tratamento de dados nas composições analíticas da GOINFRA. No entanto, optou-se por 

### c. Composições sintéticas SINAPI <a id="handle_sinapi_sintetica"></a>

#### Segmentação de colunas

Apenas as colunas de código, descrição, unidade e custo total são relevantes para as composições sintéticas.

In [22]:
sinapi_sintetica = sinapi_analitica[['CODIGO DA COMPOSICAO', 'DESCRICAO DA COMPOSICAO', 'UNIDADE', 'CUSTO TOTAL']].dropna()
sinapi_sintetica

CODIGO DA COMPOSICAO                            DESCRICAO DA COMPOSICAO  \
1                    97141  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
2                    97141  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
3                    97141  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
4                    97141  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
5                    97141  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
...                    ...                                                ...   
42047               101460           VIGIA DIURNO COM ENCARGOS COMPLEMENTARES   
42048               101460           VIGIA DIURNO COM ENCARGOS COMPLEMENTARES   
42049               101460           VIGIA DIURNO COM ENCARGOS COMPLEMENTARES   
42050               101460           VIGIA DIURNO COM ENCARGOS COMPLEMENTARES   
42051               101460           VIGIA DIURNO COM ENCARGOS COMPLEMENTARES   

      UNIDADE CUSTO TOTAL  
1           M        6,84  
2           M        6,84  
3           M        6,84  
4           M        6,84  
5           M        6,84  
...       ...         ...  
42047     MES    2.520,79  
42048     MES    2.520,79  
42049     MES    2.520,79  
42050     MES    2.520,79  
42051     MES    2.520,79  

[42051 rows x 4 columns]

A tabela acima indica que foi efetuada uma segmentação dos dados. No entanto, há diversos dados duplicados que precisam ser removidos.

#### Eliminação de dados duplicados

A quantidade de itens únicos é apresentada abaixo.

In [23]:
len(sinapi_sintetica['CODIGO DA COMPOSICAO'].unique())

6813

Há um total de 6813 códigos únicos para composições sintéticas.

Os valores duplicados são removidos a seguir.

In [24]:
sinapi_sintetica = sinapi_sintetica.dropna().drop_duplicates().reset_index(drop=True)
#sinapi_sintetica['CUSTO TOTAL'] = sinapi_sintetica['CUSTO TOTAL'].apply(lambda x: x.replace(',', '.'))
sinapi_sintetica

CODIGO DA COMPOSICAO                            DESCRICAO DA COMPOSICAO  \
0                   97141  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
1                   97142  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
2                   97143  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
3                   97144  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
4                   97145  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
...                   ...                                                ...   
6808               101456  TÉCNICO DE LABORATÓRIO E CAMPO DE CONSTRUÇÃO C...   
6809               101457    TÉCNICO EM SONDAGEM COM ENCARGOS COMPLEMENTARES   
6810               101458               TELHADOR COM ENCARGOS COMPLEMENTARES   
6811               101459             VIDRACEIRO COM ENCARGOS COMPLEMENTARES   
6812               101460           VIGIA DIURNO COM ENCARGOS COMPLEMENTARES   

     UNIDADE CUSTO TOTAL  
0          M        6,84  
1          M        7,64  
2          M        9,64  
3          M       11,64  
4          M       13,66  
...      ...         ...  
6808     MES    4.574,30  
6809     MES    5.107,62  
6810     MES    3.677,78  
6811     MES    3.248,19  
6812     MES    2.520,79  

[6813 rows x 4 columns]

In [25]:
sinapi_sintetica.isna().sum()

CODIGO DA COMPOSICAO       0
DESCRICAO DA COMPOSICAO    0
UNIDADE                    0
CUSTO TOTAL                0
dtype: int64

Após a eliminação dos dados duplicados, os valores nulos também foram removidos.

#### Conversão de tipo

Os tipos para as composições sintéticas são apresentados abaixo.

In [26]:
sinapi_sintetica.dtypes

CODIGO DA COMPOSICAO       object
DESCRICAO DA COMPOSICAO    object
UNIDADE                    object
CUSTO TOTAL                object
dtype: object

Optou-se pela conversão da coluna de custo para float.

In [27]:
sinapi_sintetica['CUSTO TOTAL'] = helpers.convert_decimal_separator(sinapi_sintetica['CUSTO TOTAL'])
sinapi_sintetica

CODIGO DA COMPOSICAO                            DESCRICAO DA COMPOSICAO  \
0                   97141  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
1                   97142  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
2                   97143  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
3                   97144  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
4                   97145  ASSENTAMENTO DE TUBO DE FERRO FUNDIDO PARA RED...   
...                   ...                                                ...   
6808               101456  TÉCNICO DE LABORATÓRIO E CAMPO DE CONSTRUÇÃO C...   
6809               101457    TÉCNICO EM SONDAGEM COM ENCARGOS COMPLEMENTARES   
6810               101458               TELHADOR COM ENCARGOS COMPLEMENTARES   
6811               101459             VIDRACEIRO COM ENCARGOS COMPLEMENTARES   
6812               101460           VIGIA DIURNO COM ENCARGOS COMPLEMENTARES   

     UNIDADE  CUSTO TOTAL  
0          M         6.84  
1          M         7.64  
2          M         9.64  
3          M        11.64  
4          M        13.66  
...      ...          ...  
6808     MES      4574.30  
6809     MES      5107.62  
6810     MES      3677.78  
6811     MES      3248.19  
6812     MES      2520.79  

[6813 rows x 4 columns]

In [28]:
sinapi_sintetica.dtypes

CODIGO DA COMPOSICAO        object
DESCRICAO DA COMPOSICAO     object
UNIDADE                     object
CUSTO TOTAL                float64
dtype: object

Constata-se acima que a conversão da coluna de custo foi bem sucedida.

### d. Composições analíticas SINAPI <a id="handle_sinapi_analitica"></a>

##### Segmentação de colunas

Apenas as colunas de códigos, descrição, unidade, coeficiente, preço unitário e custo total são relevantes para nossa análise.

In [29]:
sinapi_analitica = sinapi_analitica[['CODIGO DA COMPOSICAO', 'CODIGO ITEM', 'DESCRIÇÃO ITEM', 'UNIDADE ITEM', 'COEFICIENTE', 'PRECO UNITARIO', 'CUSTO TOTAL.1']].reset_index(drop=True)
sinapi_analitica

CODIGO DA COMPOSICAO  CODIGO ITEM  \
0                      NaN          NaN   
1                    97141          NaN   
2                    97141       5678.0   
3                    97141       5679.0   
4                    97141      20078.0   
...                    ...          ...   
42047               101460      40864.0   
42048               101460      41096.0   
42049               101460      43479.0   
42050               101460      43503.0   
42051               101460     101372.0   

                                          DESCRIÇÃO ITEM UNIDADE ITEM  \
0                                                    NaN          NaN   
1                                                    NaN          NaN   
2      RETROESCAVADEIRA SOBRE RODAS COM CARREGADEIRA,...          CHP   
3      RETROESCAVADEIRA SOBRE RODAS COM CARREGADEIRA,...          CHI   
4      PASTA LUBRIFICANTE PARA TUBOS E CONEXOES COM J...           UN   
...                                                  ...          ...   
42047               SEGURO - MENSALISTA (COLETADO CAIXA)          MES   
42048                          VIGIA DIURNO (MENSALISTA)          MES   
42049  FERRAMENTAS - FAMILIA SERVENTE - MENSALISTA (E...          MES   
42050  EPI - FAMILIA SERVENTE - MENSALISTA (ENCARGOS ...          MES   
42051  CURSO DE CAPACITAÇÃO PARA VIGIA DIURNO (ENCARG...          MES   

      COEFICIENTE PRECO UNITARIO CUSTO TOTAL.1  
0             NaN            NaN           NaN  
1             NaN            NaN           NaN  
2       0,0099000          97,30          0,96  
3       0,0477000          45,41          2,16  
4       0,0046000          23,15          0,10  
...           ...            ...           ...  
42047   1,0000000          11,13         11,13  
42048   1,0000000       1.673,15      1.673,15  
42049   1,0000000          76,97         76,97  
42050   1,0000000         191,25        191,25  
42051   1,0000000           4,68          4,68  

[42052 rows x 7 columns]

##### Eliminação de dados nulos

O número de itens nulos é apresentado a seguir.

In [30]:
sinapi_analitica.isna().sum()

CODIGO DA COMPOSICAO       1
CODIGO ITEM             6814
DESCRIÇÃO ITEM          6814
UNIDADE ITEM            6814
COEFICIENTE             6814
PRECO UNITARIO          6814
CUSTO TOTAL.1           6814
dtype: int64

É possível afirmar que a segmentação de colunas eliminou os dados nulos. Portanto, não há necessidade de tratá-los novamente.

##### Conversão de tipo

As colunas preço unitário e custo total devem ser convertidas para float.

In [31]:
sinapi_analitica.dtypes

CODIGO DA COMPOSICAO     object
CODIGO ITEM             float64
DESCRIÇÃO ITEM           object
UNIDADE ITEM             object
COEFICIENTE              object
PRECO UNITARIO           object
CUSTO TOTAL.1            object
dtype: object

A conversão é feita logo abaixo.

In [32]:
sinapi_analitica['PRECO UNITARIO'] = helpers.convert_decimal_separator(sinapi_analitica['PRECO UNITARIO'])
sinapi_analitica['CUSTO TOTAL.1'] = helpers.convert_decimal_separator(sinapi_analitica['CUSTO TOTAL.1'])

AttributeError: 'float' object has no attribute 'replace'

Após a conversão, os tipos de dados são indicados.

In [ ]:
sinapi_analitica.dtypes

Percebe-se que a conversão foi bem-sucedida.

In [ ]:
sinapi_analitica.loc[sinapi_analitica['CODIGO DA COMPOSICAO'] == '97141']

In [ ]:
teste = pd.merge(goinfra_sintetica, goinfra_analitica, on=['CODIGO DA COMPOSICAO'])
teste.drop(columns=['UNIDADE', 'CUSTO MATERIAL', 'CUSTO MAO DE OBRA', 'CUSTO TOTAL'])

In [ ]:
sintetica = data[['CODIGO DA COMPOSICAO', 'DESCRICAO DA COMPOSICAO', 'UNIDADE', 'CUSTO TOTAL']]
sintetica